### PR TITLE
gh-101100: Fix Sphinx warning in gc.rst and refactor docs clean list

### PR DESCRIPTION
**Pull request by hugovk**
Fix the single Sphinx warning in `gc.rst`, add the file to the clean list (so new warnings will fail the build), and refactor the clean list into a little Python script to keep the GitHub workflow from getting too big.

* Issue: gh-101100

### CODE REVIEW

File: Doc/tools/touch-clean-files.py
Comment (CAM-Gerlach) [inline Doc/tools/touch-clean-files.py]: Since these are already strings, why not

```suggestion
    print(filename)
```

File: Doc/tools/touch-clean-files.py
Comment (CAM-Gerlach) [inline Doc/tools/touch-clean-files.py]: I know I mentioned this before, but it would be really nice to have this as a static JSON (etc.) data file, or even a simple text file with a blank line between entries (to reduce merge conflicts), rather than part of the dynamic executable code here...

**Review by CAM-Gerlach — COMMENTED** [review]
A couple comments, otherwise LGTM, thanks

File: Doc/tools/touch-clean-files.py
Comment (hugovk) [reply Doc/tools/touch-clean-files.py]: Of course :) I started off with "Touching " in there too but forgot to de-eff it.

File: Doc/tools/touch-clean-files.py
Comment (hugovk) [reply Doc/tools/touch-clean-files.py]: What's the benefit of moving it to a static data file? Seems like extra complexity for a simple script, and we're not using the data elsewhere. We could have blank lines in here to avoid conflicts.

Comment (CAM-Gerlach) [reply Doc/tools/touch-clean-files.py]: IMO, but:

* Its cleaner to separate the code and data (especially once the latter dominates)
* Is more concise to write as the length of this list starts to dominate the code in the file (especially if in a simple .gitginore style text format—no indent, no quotes, no comma, etc)
* Is much more more readable/consumable by any other tooling we or others may introduce down the road
* Avoids more complex history and merge conflicts if we ever move it later

None of these are huge reasons on their own, but it seems worth it overall compared to the relatively extra complexity of reading in a gitignore-style simple text file:

```python
CLEAN = Path("clean.txt").read_text(encoding="UTF-8").split("\n")
```

File: Doc/tools/touch-clean-files.py
Comment (CAM-Gerlach) [reply Doc/tools/touch-clean-files.py]: Hehe, figured it was something like that

File: Doc/tools/touch-clean-files.py
Comment (hugovk) [reply Doc/tools/touch-clean-files.py]: OK! How's it now?

File: Doc/tools/touch-clean-files.py
Comment (CAM-Gerlach) [inline Doc/tools/touch-clean-files.py]:
```suggestion
    if filename.strip():
        print(filename)
        Path(filename.strip()).touch()
```

Avoid any issues with trailing (or leading) whitespace, either on empty or non-empty lines.

**Review by CAM-Gerlach — COMMENTED** [review]
One followup comment, otherwise LGTM — thanks @hugovk !

File: Doc/tools/touch-clean-files.py
Comment (CAM-Gerlach) [reply Doc/tools/touch-clean-files.py]: Yup, looks great aside from one followup comment :+1: Thanks!

File: Doc/tools/clean-files.txt
Comment (ezio-melotti) [inline Doc/tools/clean-files.txt:10]: I don't want to add too much feature creep in the PR, but it would be useful to add a brief comment at the top to say:
1. what the file is for and who uses it;
2. why there are blank lines.

The code can then be updated to something like this to ignore the comments:
```py
with Path("Doc/tools/clean-files.txt").open() as clean_files:
    CLEAN = [fname for fname in clean_files
             if fname.strip() and not fname.startswith('#')]
```

File: Doc/tools/touch-clean-files.py
Comment (ezio-melotti) [reply Doc/tools/touch-clean-files.py]: This could be walrused to avoid duplication, or if my previous suggestion is accepted it could be added to `CLEAN = [fname.strip() ...]` directly and both the `if` and the `strip` removed.

Another option is to avoid the `strip` and let it fail if there is extra whitespace so that the user can remove it.  If we do this, we should provide an easy way for the user to test this locally before creating the PR, either as part of `make patchcheck` (is it still used?), or as a target in the `Doc`'s `makefile`.  This can be done in a separate PR though.

File: Doc/tools/clean-files.txt
Comment (CAM-Gerlach) [reply Doc/tools/clean-files.txt:10]: Yup, I was thinking of requesting the same, but I felt I'd already asked for enough, heh

File: Doc/tools/touch-clean-files.py
Comment (CAM-Gerlach) [reply Doc/tools/touch-clean-files.py]: > if my previous suggestion is accepted it could be added to `CLEAN = [fname.strip() ...]` directly and both the `if` and the `strip` removed.

+1 to that approach

> Another option is to avoid the `strip` and let it fail if there is extra whitespace so that the user can remove it.

As much as I'm pretty nitpicky, this seems like being a bit too much of a stickler for even my taste to fail the whole CIs for considering its trivial to ignore, especially considering the check warnings are already eating into our core dev annoyance budget with this :)

File: Doc/tools/clean-files.txt
Comment (hugovk) [reply Doc/tools/clean-files.txt:10]: Done!

File: Doc/tools/clean-files.txt
Comment (CAM-Gerlach) [inline Doc/tools/clean-files.txt]:
```suggestion
# Add blank lines between files and keep them sorted lexicographically
# to help avoid merge conflicts.
```

Keeping it at least roughly sorted is important not only for checking if something's already included, but also avoids inevitable merge conflicts if things are always added at the bottom.

If we were using pre-commit (or if we switch to it at some point), we could use a hook for that, of course.

File: Doc/tools/touch-clean-files.py
Comment (CAM-Gerlach) [inline Doc/tools/touch-clean-files.py]:
```suggestion
        Path(filename.strip())
```

May as well just do the Path conversion here, eh?

File: Doc/tools/touch-clean-files.py
Comment (CAM-Gerlach) [inline Doc/tools/touch-clean-files.py]:
```suggestion
    print(filename)
    filename.touch()
```

* You're already checking `if filename` above, so it can be removed here
* Also, I moved the `Path()` coercion above, with the rest

**Review by CAM-Gerlach — COMMENTED** [review]
Sorry, a couple more followup fixes after @ezio-melotti 's suggested changes...

**Comment by hugovk** [comment]
Thanks! I've got a list of ~180 other files we can already add to `clean-files.txt`. Shall we do it in this PR?

**Comment by CAM-Gerlach** [comment]
My inclination would be to leave that for a followup PR, but I don't feel strongly about it and would of course defer to you and @ezio-melotti if you do.

**Review by CAM-Gerlach — APPROVED** [review]
LGTM, thanks @hugovk !

**Comment by AlexWaygood** [comment]
(Doesn't need to be done in this PR.)

Maybe it would be better to keep a "dirty list" rather than a "clean list". I'd quite like to help out with getting the number of sphinx warnings down, and it would be handy to have a guaranteed-to-be-up-to-date list of all files that have yet to be worked on. Using a dirty list rather than a clean list would also mean that any new files added to the `Doc/` directory would by default be tested with nitpicks enabled, whereas currently the default is for new files to be tested without nitpicks enabled.

The only downside I can see is that the dirty list would have to initially be pretty long. But I don't think that's actually a _problem_; it would be helpful to have an easy visualisation of the scale of the task still to be done.

**Comment by CAM-Gerlach** [comment]
Yeah, that sounds like a pretty compelling to me for the aforementioned followup. If we are going to take that approach, which certainly might make sense and matches the pattern of listing files to ignore rather than check (`.gitignore`, etc). In that event, we would probably want to specify directory-level exclude support, as well as an include list so we could, e.g, exclude all the files in `Doc/whatsnew` except for the current version's, without having to manually set and update that for every branch, to minimize ongoing maintenance (since that would be permanent).

**Comment by AlexWaygood** [comment]
> In that event, we would probably want to specify directory-level exclude support, as well as an include list so we could, e.g, exclude all the files in `Doc/whatsnew` except for the current version's, without having to manually set and update that for every branch, to minimize ongoing maintenance (since that would be permanent).

You could add regex or glob support for the excludelist, but I'd personally vote against supporting an includelist and an excludelist simultaneously. That sounds like it would be overcomplicating things unnecessarily imho. Just add all files in `Doc/whatsnew` except `3.12.rst` to the excludelist for now, and remove them one by one as we sort them out (or add a comment next to that section of the excludelist saying that it's not high priority to fix them).

We've done a very similar thing for a while for our stubtest check over at typeshed. Stubtest is a script that verifies that various things in the stub are consistent in various ways with the objects at runtime. It's a very useful tool, but it necessitates keeping a long allowlist -- some of these are inconsistencies that are still to be completed, but many have comments next to them or above them indicating that they can't be resolved for one reason or another. The system works very well for us. (Stubtest's allowlist has regex support, but not globbing support.)

https://github.com/python/typeshed/blob/main/tests/stubtest_allowlists/py3_common.txt

**Comment by hugovk** [comment]
> I'd quite like to help out with getting the number of sphinx warnings down

Great!

> The only downside I can see is that the dirty list would have to initially be pretty long. But I don't think that's actually a _problem_; it would be helpful to have an easy visualisation of the scale of the task still to be done.

Sure. There's ~490 total RST files, ~180 clean, so ~310 dirty. Not such a big difference.

Given a dirty list, we'd have to compute the clean list, so we can `touch` them to rebuild them.

I made the initial clean list by subtracting the list of files with warnings from a list of all files. I tested with all of those, but found that `touch`ing two clean files (`includes/wasm-notavail.rst` and `whatsnew/changelog.rst`) caused a cascade effect and resulted in a number of dirty files being rebuilt too, and so failing the build. So I've removed those. We'd need a mechanism to exclude them, and potentially others.

**Comment by hugovk** [comment]
> My inclination would be to leave that for an immediate followup PR to keep them separate in the commit history and avoid lengthening this one further, but I don't feel strongly about it and would of course defer to you and @ezio-melotti if you do.
> 
> It would be good to do as an immediate follow-up, though, as it will help avoid regressions and also merge conflicts when adding more files (which are far more likely with a short file list). Thanks!

Sure, I'll merge this now and make a new PR right away!

**Comment by hugovk** [comment]
➡️ https://github.com/python/cpython/pull/103135

**Comment by hugovk** [comment]
> (Doesn't need to be done in this PR.)
> 
> Maybe it would be better to keep a "dirty list" rather than a "clean list". I'd quite like to help out with getting the number of sphinx warnings down, and it would be handy to have a guaranteed-to-be-up-to-date list of all files that have yet to be worked on. Using a dirty list rather than a clean list would also mean that any new files added to the `Doc/` directory would by default be tested with nitpicks enabled, whereas currently the default is for new files to be tested without nitpicks enabled.
> 
> The only downside I can see is that the dirty list would have to initially be pretty long. But I don't think that's actually a _problem_; it would be helpful to have an easy visualisation of the scale of the task still to be done.

Please see GH-103191.

**Pull request merged**
